File: whatsapp-mcp-server/lib/identity.py
"""Read access to the unified identity directory (the `identities` table).

The bridge maintains the table; everything here is read-only. It collapses what
WhatsApp splits: the same person arrives under a phone JID on one route and a
hidden `@lid` on another, and whatsmeow's contact store keeps a separate row for
each — usually the address-book name on one and the self-chosen push name on the
other. A caller holding either identity gets the whole record in one lookup.

Every function degrades to an empty result when the table is missing or empty
(a store that predates the directory, or a bridge that has not synced yet), so
callers can fall back to the older per-source queries.
"""

import sqlite3
from dataclasses import dataclass, field
from typing import Any

from lib.utils import MESSAGES_DB_PATH, logger

IdentityKind = str  # user | group | newsletter | broadcast

_COLUMNS = """
    jid, lid, phone_jid, phone, redacted_phone, kind,
    first_name, full_name, push_name, business_name, nickname, display_name, chat_name,
    is_business, is_contact, has_chat, message_count,
    participant_count, is_announce, owner_jid,
    first_seen, last_seen, sources, updated_at
"""


@dataclass
class Identity:
    """One directory row: a person, group, newsletter or broadcast list."""

    jid: str
    display_name: str
    kind: IdentityKind = "user"
    lid: str | None = None
    phone_jid: str | None = None
    phone: str | None = None
    redacted_phone: str | None = None
    first_name: str | None = None
    full_name: str | None = None
    push_name: str | None = None
    business_name: str | None = None
    nickname: str | None = None
    chat_name: str | None = None
    is_business: bool = False
    is_contact: bool = False
    has_chat: bool = False
    message_count: int = 0
    participant_count: int | None = None
    is_announce: bool = False
    owner_jid: str | None = None
    first_seen: str | None = None
    last_seen: str | None = None
    sources: list[str] = field(default_factory=list)
    updated_at: str | None = None

    def to_dict(self) -> dict[str, Any]:
        """Convert to a dictionary for structured tool output.

        Group-only and person-only fields are omitted rather than emitted as
        nulls — the directory holds both kinds, and a caller reading a group
        should not have to skip past six empty name fields.
        """
        result: dict[str, Any] = {
            "jid": self.jid,
            "display_name": self.display_name,
            "kind": self.kind,
            "has_chat": self.has_chat,
            "message_count": self.message_count,
            "last_seen": self.last_seen,
        }
        if self.kind == "group":
            result.update(
                {
                    "participant_count": self.participant_count,
                    "is_announce": self.is_announce,
                    "owner_jid": self.owner_jid,
                }
            )
        else:
            result.update(
                {
                    "lid": self.lid,
                    "phone_jid": self.phone_jid,
                    "phone": self.phone,
                    "is_contact": self.is_contact,
                    "is_business": self.is_business,
                    "names": {
                        "nickname": self.nickname,
                        "full_name": self.full_name,
                        "first_name": self.first_name,
                        "push_name": self.push_name,
                        "business_name": self.business_name,
                        "chat_name": self.chat_name,
                    },
                }
            )
            if self.redacted_phone and not self.phone:
                result["redacted_phone"] = self.redacted_phone
        result["first_seen"] = self.first_seen
        result["sources"] = self.sources
        result["updated_at"] = self.updated_at
        return result


def _row_to_identity(row: tuple) -> Identity:
    return Identity(
        jid=row[0],
        lid=row[1],
        phone_jid=row[2],
        phone=row[3],
        redacted_phone=row[4],
        kind=row[5] or "user",
        first_name=row[6],
        full_name=row[7],
        push_name=row[8],
        business_name=row[9],
        nickname=row[10],
        display_name=row[11] or row[0],
        chat_name=row[12],
        is_business=bool(row[13]),
        is_contact=bool(row[14]),
        has_chat=bool(row[15]),
        message_count=row[16] or 0,
        participant_count=row[17],
        is_announce=bool(row[18]),
        owner_jid=row[19],
        first_seen=row[20],
        last_seen=row[21],
        sources=[s for s in (row[22] or "").split(",") if s],
        updated_at=row[23],
    )


def directory_available(conn: sqlite3.Connection | None = None) -> bool:
    """Whether the directory exists and has been populated at least once.

    Checked per call rather than cached: the bridge can create and fill the
    table while a long-lived MCP session is running, and a cached False would
    pin that session to the fallback path for its whole lifetime.
    """
    own_conn = conn is None
    try:
        conn = conn or sqlite3.connect(MESSAGES_DB_PATH)
        cursor = conn.execute("SELECT name FROM sqlite_master WHERE type = 'table' AND name = 'identities'")
        if cursor.fetchone() is None:
            return False
        return conn.execute("SELECT 1 FROM identities LIMIT 1").fetchone() is not None
    except sqlite3.Error as e:
        logger.debug(f"Identity directory unavailable: {e}")
        return False
    finally:
        if own_conn and conn is not None:
            conn.close()


def get_identity(value: str) -> Identity | None:
    """Look up one identity by any handle we know it under.

    Accepts a canonical JID, a LID, a phone JID, a bare phone number, or an
    exact display name. Anything a message, a chat row or a human is likely to
    be holding resolves to the same record.
    """
    if not value or not value.strip():
        return None
    value = value.strip()
    bare = value.split("@")[0]

    try:
        conn = sqlite3.connect(MESSAGES_DB_PATH)
        if not directory_available(conn):
            return None
        cursor = conn.execute(
            f"""
            SELECT {_COLUMNS} FROM identities
            WHERE jid = ? OR lid = ? OR phone_jid = ? OR phone = ?
               OR lid = ? OR phone_jid = ?
               OR LOWER(display_name) = LOWER(?)
            ORDER BY CASE WHEN jid = ? THEN 0 ELSE 1 END, message_count DESC
            LIMIT 1
            """,
            (
                value,
                value,
                value,
                bare,
                f"{bare}@lid",
                f"{bare}@s.whatsapp.net",
                value,
                value,
            ),
        )
        row = cursor.fetchone()
        return _row_to_identity(row) if row else None
    except sqlite3.Error as e:
        logger.error(f"Database error in get_identity: {e}")
        return None
    finally:
        if "conn" in locals():
            conn.close()


def resolve_jid(value: str) -> str | None:
    """Return the canonical JID for any handle, or None if unknown."""
    identity = get_identity(value)
    return identity.jid if identity else None


def list_identities(
    query: str | None = None,
    kind: str | None = None,
    has_chat: bool | None = None,
    is_contact: bool | None = None,
    sort_by: str = "last_active",
    limit: int = 50,
    page: int = 0,
) -> list[Identity]:
    """Search the directory.

    `query` matches any name, any JID form, and the phone number, so a caller
    can pass whatever fragment it has without knowing which field it belongs to.
    """
    try:
        conn = sqlite3.connect(MESSAGES_DB_PATH)
        if not directory_available(conn):
            return []

        where: list[str] = []
        params: list[Any] = []

        if query:
            pattern = f"%{query.strip()}%"
            where.append(
                """(
                    LOWER(COALESCE(display_name, '')) LIKE LOWER(?)
                    OR LOWER(COALESCE(full_name, '')) LIKE LOWER(?)
                    OR LOWER(COALESCE(first_name, '')) LIKE LOWER(?)
                    OR LOWER(COALESCE(push_name, '')) LIKE LOWER(?)
                    OR LOWER(COALESCE(business_name, '')) LIKE LOWER(?)
                    OR LOWER(COALESCE(nickname, '')) LIKE LOWER(?)
                    OR LOWER(COALESCE(chat_name, '')) LIKE LOWER(?)
                    OR jid LIKE ? OR COALESCE(lid, '') LIKE ? OR COALESCE(phone, '') LIKE ?
                )"""
            )
            params.extend([pattern] * 10)

        if kind:
            where.append("kind = ?")
            params.append(kind)
        if has_chat is not None:
            where.append("has_chat = ?")
            params.append(1 if has_chat else 0)
        if is_contact is not None:
            where.append("is_contact = ?")
            params.append(1 if is_contact else 0)

        order = {
            "last_active": "last_seen DESC, message_count DESC",
            "name": "display_name COLLATE NOCASE",
            "messages": "message_count DESC",
        }.get(sort_by, "last_seen DESC, message_count DESC")

        sql = f"SELECT {_COLUMNS} FROM identities"
        if where:
            sql += " WHERE " + " AND ".join(where)
        sql += f" ORDER BY {order} LIMIT ? OFFSET ?"
        params.extend([limit, page * limit])

        return [_row_to_identity(row) for row in conn.execute(sql, tuple(params)).fetchall()]
    except sqlite3.Error as e:
        logger.error(f"Database error in list_identities: {e}")
        return []
    finally:
        if "conn" in locals():
            conn.close()


def directory_stats() -> dict[str, Any]:
    """Coverage summary — how much of the directory is actually resolved.

    Useful on its own (how many contacts do I have?) and as the answer to "why
    is this name still a phone number?": a low `linked` count means the bridge
    has not seen the LID mapping for those people yet.
    """
    try:
        conn = sqlite3.connect(MESSAGES_DB_PATH)
        if not directory_available(conn):
            return {"available": False, "reason": "identities table missing or not yet synced by the bridge"}

        row = conn.execute(
            """
            SELECT
                COUNT(*),
                SUM(kind = 'user'),
                SUM(kind = 'group'),
                SUM(lid IS NOT NULL AND phone_jid IS NOT NULL),
                SUM(lid IS NOT NULL AND phone_jid IS NULL),
                SUM(is_contact = 1),
                SUM(has_chat = 1),
                MAX(updated_at)
            FROM identities
            """
        ).fetchone()

        # merged_into arrives with the same migration as the directory, but a
        # store can be half-migrated; a missing column is not worth failing on.
        try:
            merged = conn.execute("SELECT COUNT(*) FROM chats WHERE merged_into IS NOT NULL").fetchone()[0]
        except sqlite3.Error:
            merged = None

        return {
            "available": True,
            "total": row[0],
            "users": row[1] or 0,
            "groups": row[2] or 0,
            "lid_and_phone_known": row[3] or 0,
            "lid_only": row[4] or 0,
            "in_address_book": row[5] or 0,
            "with_chat": row[6] or 0,
            "merged_lid_chats": merged,
            "last_synced": row[7],
        }
    except sqlite3.Error as e:
        logger.error(f"Database error in directory_stats: {e}")
        return {"available": False, "reason": str(e)}
    finally:
        if "conn" in locals():
            conn.close()

File: whatsapp-bridge/internal/whatsapp/identity.go
package whatsapp

import (
	"context"
	"fmt"
	"sort"
	"strings"
	"time"

	"go.mau.fi/whatsmeow/types"
	"whatsapp-bridge/internal/database"
)

// IdentitySyncStats summarises one run of SyncIdentities.
type IdentitySyncStats struct {
	Rows        int
	Users       int
	Groups      int
	Linked      int // rows where both a LID and a phone JID are known
	Pruned      int
	GroupsEnric int // groups enriched with live metadata from the server
	Duration    time.Duration
}

// LIDMergeStats summarises one run of MergeLIDChats.
type LIDMergeStats struct {
	ChatsMerged   int
	MessagesMoved int
	Duplicates    int
	Unresolved    int // LID chats with no phone mapping yet — retried next run
}

// SyncIdentities rebuilds the unified directory in messages.db.
//
// Everything it needs already exists, just scattered: whatsmeow's contact store
// keys names by JID and files the same person twice (once under the phone JID
// with the address-book name, once under the LID with the push name), its LID
// map knows which two are the same person but is joined to nothing, and the
// chat and message tables know who we actually talk to. This collapses all of
// it onto one canonical JID per person or group.
//
// Safe to run repeatedly; it is a full rebuild, not an increment.
func (c *Client) SyncIdentities(messageStore *database.MessageStore) (IdentitySyncStats, error) {
	started := time.Now()
	var stats IdentitySyncStats
	ctx := context.Background()

	byJID := make(map[string]*database.Identity)

	// Contacts first, phone-JID rows before LID rows. Field merging is
	// fill-if-empty, so this ordering is what makes the address-book name win
	// over the self-chosen push name — and what makes the result deterministic
	// rather than dependent on Go's map iteration order.
	contacts, err := c.Store.Contacts.GetAllContacts(ctx)
	if err != nil {
		return stats, fmt.Errorf("read contact store: %w", err)
	}

	contactJIDs := make([]types.JID, 0, len(contacts))
	for jid := range contacts {
		contactJIDs = append(contactJIDs, jid)
	}
	sort.Slice(contactJIDs, func(i, j int) bool {
		a, b := contactJIDs[i], contactJIDs[j]
		if (a.Server == types.DefaultUserServer) != (b.Server == types.DefaultUserServer) {
			return a.Server == types.DefaultUserServer
		}
		return a.String() < b.String()
	})

	for _, jid := range contactJIDs {
		info := contacts[jid]
		row := c.identityFor(ctx, byJID, jid)
		if row == nil {
			continue
		}
		fillIfEmpty(&row.FirstName, info.FirstName)
		fillIfEmpty(&row.FullName, info.FullName)
		fillIfEmpty(&row.PushName, info.PushName)
		fillIfEmpty(&row.BusinessName, info.BusinessName)
		fillIfEmpty(&row.RedactedPhone, info.RedactedPhone)
		if info.BusinessName != "" {
			row.IsBusiness = true
		}
		if info.Found {
			row.IsContact = true
		}
		addSource(row, "contacts")
	}

	// Chats contribute groups, newsletters, and anyone we talk to who never made
	// it into the address book.
	chatRows, err := messageStore.GetChatRows()
	if err != nil {
		return stats, fmt.Errorf("read chats: %w", err)
	}
	for _, chat := range chatRows {
		jid, err := types.ParseJID(chat.JID)
		if err != nil {
			c.logger.Debugf("[IDENTITY] skipping unparseable chat jid %q: %v", chat.JID, err)
			continue
		}
		row := c.identityFor(ctx, byJID, jid)
		if row == nil {
			continue
		}
		// A merged-away chat still tells us the identity exists, but its
		// activity has already been folded into the target's row.
		if chat.MergedInto == "" {
			row.HasChat = true
			fillIfEmpty(&row.ChatName, chat.Name)
			row.LastSeen = laterOf(row.LastSeen, chat.LastMessageTime)
		}
		addSource(row, "chats")
	}

	// Message aggregates, folded onto the canonical identity so a conversation
	// that was split across two JIDs counts once.
	chatStats, err := messageStore.GetChatStats()
	if err != nil {
		return stats, fmt.Errorf("read chat stats: %w", err)
	}
	for chatJID, stat := range chatStats {
		jid, err := types.ParseJID(chatJID)
		if err != nil {
			continue
		}
		row := c.identityFor(ctx, byJID, jid)
		if row == nil {
			continue
		}
		row.MessageCount += stat.Count
		row.FirstSeen = earlierOf(row.FirstSeen, stat.FirstSeen)
		row.LastSeen = laterOf(row.LastSeen, stat.LastSeen)
		addSource(row, "messages")
	}

	// Nicknames are a deliberate user override and win the display name.
	nicknames, err := messageStore.GetNicknames()
	if err != nil {
		return stats, fmt.Errorf("read nicknames: %w", err)
	}
	for nickJID, nickname := range nicknames {
		jid, err := types.ParseJID(nickJID)
		if err != nil {
			continue
		}
		row := c.identityFor(ctx, byJID, jid)
		if row == nil {
			continue
		}
		row.Nickname = nickname
		addSource(row, "nicknames")
	}

	stats.GroupsEnric = c.enrichGroups(ctx, byJID)

	rows := make([]database.Identity, 0, len(byJID))
	keep := make([]string, 0, len(byJID))
	for _, row := range byJID {
		row.DisplayName = row.ResolveDisplayName()
		sort.Strings(row.Sources)
		rows = append(rows, *row)
		keep = append(keep, row.JID)
		switch row.Kind {
		case "group":
			stats.Groups++
		case "user":
			stats.Users++
		}
		if row.LID != "" && row.PhoneJID != "" {
			stats.Linked++
		}
	}

	if err := messageStore.UpsertIdentities(rows); err != nil {
		return stats, err
	}
	pruned, err := messageStore.PruneIdentities(keep)
	if err != nil {
		return stats, err
	}

	stats.Rows = len(rows)
	stats.Pruned = pruned
	stats.Duration = time.Since(started)
	return stats, nil
}

// identityFor returns the directory row a JID belongs to, creating it on first
// sight. Both halves of a split identity resolve to the same row.
func (c *Client) identityFor(ctx context.Context, byJID map[string]*database.Identity, jid types.JID) *database.Identity {
	jid = jid.ToNonAD()
	canonical, lid, phone := c.resolveIdentity(ctx, jid)
	if canonical.IsEmpty() {
		return nil
	}

	key := canonical.String()
	row, ok := byJID[key]
	if !ok {
		row = &database.Identity{JID: key, Kind: kindForServer(canonical.Server)}
		byJID[key] = row
	}
	if !lid.IsEmpty() {
		row.LID = lid.String()
	}
	if !phone.IsEmpty() {
		row.PhoneJID = phone.String()
		row.Phone = phone.User
	}
	return row
}

// resolveIdentity maps any JID onto its canonical form plus both known halves.
// The phone JID is canonical whenever we know it: it is stable, human-readable,
// and is what every other tool in this repo already keys on.
func (c *Client) resolveIdentity(ctx context.Context, jid types.JID) (canonical, lid, phone types.JID) {
	switch jid.Server {
	case types.HiddenUserServer:
		lid = jid
		if resolved, err := c.Store.GetAltJID(ctx, jid); err == nil && !resolved.IsEmpty() {
			phone = resolved.ToNonAD()
		}
	case types.DefaultUserServer:
		phone = jid
		if resolved, err := c.Store.GetAltJID(ctx, jid); err == nil && !resolved.IsEmpty() {
			lid = resolved.ToNonAD()
		}
	default:
		return jid, types.EmptyJID, types.EmptyJID
	}

	if !phone.IsEmpty() {
		return phone, lid, phone
	}
	return jid, lid, phone
}

// enrichGroups pulls live group metadata in a single round trip. Best-effort:
// the directory is still useful without it, so a failure is logged, not fatal.
func (c *Client) enrichGroups(ctx context.Context, byJID map[string]*database.Identity) int {
	if !c.IsConnected() {
		return 0
	}
	groups, err := c.GetJoinedGroups(ctx)
	if err != nil {
		c.logger.Debugf("[IDENTITY] group enrichment skipped: %v", err)
		return 0
	}

	enriched := 0
	for _, group := range groups {
		row, ok := byJID[group.JID.String()]
		if !ok {
			row = &database.Identity{JID: group.JID.String(), Kind: "group"}
			byJID[row.JID] = row
		}
		if group.Name != "" {
			row.FullName = group.Name
			row.ChatName = group.Name
		}
		row.ParticipantCount = group.ParticipantCount
		if row.ParticipantCount == 0 {
			row.ParticipantCount = len(group.Participants)
		}
		row.IsAnnounce = group.IsAnnounce
		if !group.GroupCreated.IsZero() {
			created := group.GroupCreated
			row.FirstSeen = earlierOf(row.FirstSeen, &created)
		}
		if !group.OwnerPN.IsEmpty() {
			row.OwnerJID = group.OwnerPN.ToNonAD().String()
		} else if !group.OwnerJID.IsEmpty() {
			row.OwnerJID = group.OwnerJID.ToNonAD().String()
		}
		addSource(row, "groups")
		enriched++
	}
	return enriched
}

// MergeLIDChats folds every LID-filed chat into its phone-JID twin.
//
// NormalizeChatJID keeps new messages from splitting, but history written
// before it existed is already split — and a LID chat whose mapping only
// arrives later splits again until the mapping lands. Running this on each
// connect catches both.
func (c *Client) MergeLIDChats(messageStore *database.MessageStore) (LIDMergeStats, error) {
	var stats LIDMergeStats
	ctx := context.Background()

	chatRows, err := messageStore.GetChatRows()
	if err != nil {
		return stats, fmt.Errorf("read chats: %w", err)
	}

	for _, chat := range chatRows {
		if chat.MergedInto != "" {
			continue
		}
		jid, err := types.ParseJID(chat.JID)
		if err != nil || jid.Server != types.HiddenUserServer {
			continue
		}
		resolved, err := c.Store.GetAltJID(ctx, jid)
		if err != nil || resolved.IsEmpty() {
			stats.Unresolved++
			continue
		}
		target := resolved.ToNonAD().String()
		if target == chat.JID {
			continue
		}

		result, err := messageStore.MergeChat(chat.JID, target)
		if err != nil {
			c.logger.Warnf("[IDENTITY] merge %s -> %s failed: %v", chat.JID, target, err)
			continue
		}
		stats.ChatsMerged++
		stats.MessagesMoved += result.MessagesMoved
		stats.Duplicates += result.MessagesDuplicate
		c.logger.Infof("[IDENTITY] merged %s into %s (%d messages moved, %d duplicates dropped)",
			chat.JID, target, result.MessagesMoved, result.MessagesDuplicate)
	}

	return stats, nil
}

func kindForServer(server string) string {
	switch server {
	case types.GroupServer:
		return "group"
	case types.NewsletterServer:
		return "newsletter"
	case types.BroadcastServer:
		return "broadcast"
	default:
		return "user"
	}
}

func fillIfEmpty(dst *string, value string) {
	if strings.TrimSpace(*dst) == "" && strings.TrimSpace(value) != "" {
		*dst = strings.TrimSpace(value)
	}
}

func addSource(row *database.Identity, source string) {
	for _, existing := range row.Sources {
		if existing == source {
			return
		}
	}
	row.Sources = append(row.Sources, source)
}

func earlierOf(a, b *time.Time) *time.Time {
	if b == nil || b.IsZero() {
		return a
	}
	if a == nil || a.IsZero() || b.Before(*a) {
		return b
	}
	return a
}

func laterOf(a, b *time.Time) *time.Time {
	if b == nil || b.IsZero() {
		return a
	}
	if a == nil || a.IsZero() || b.After(*a) {
		return b
	}
	return a
}

// IdentitySyncer keeps the directory fresh without rebuilding it on every
// contact event. Contact, push-name and business-name events arrive in bursts
// during app-state sync — hundreds in a few seconds — and a full rebuild costs
// a table scan, so events only mark the directory dirty and a single worker
// coalesces them.
type IdentitySyncer struct {
	client   *Client
	store    *database.MessageStore
	dirty    chan struct{}
	interval time.Duration
	merge    bool
}

// NewIdentitySyncer wires a syncer to a store. Call Start to run it.
func (c *Client) NewIdentitySyncer(messageStore *database.MessageStore, interval time.Duration, merge bool) *IdentitySyncer {
	return &IdentitySyncer{
		client:   c,
		store:    messageStore,
		dirty:    make(chan struct{}, 1),
		interval: interval,
		merge:    merge,
	}
}

// Trigger requests a rebuild. Never blocks: a pending request already covers
// this one, since the rebuild is a full refresh rather than an increment.
func (s *IdentitySyncer) Trigger() {
	if s == nil {
		return
	}
	select {
	case s.dirty <- struct{}{}:
	default:
	}
}

// Start runs the sync worker until ctx is cancelled.
func (s *IdentitySyncer) Start(ctx context.Context) {
	go func() {
		var tick <-chan time.Time
		if s.interval > 0 {
			ticker := time.NewTicker(s.interval)
			defer ticker.Stop()
			tick = ticker.C
		}
		for {
			select {
			case <-ctx.Done():
				return
			case <-s.dirty:
				s.RunOnce()
			case <-tick:
				s.RunOnce()
			}
		}
	}()
}

// RunOnce merges any newly resolvable LID chats, then rebuilds the directory.
// Merge runs first so the rebuild sees the collapsed chat and counts each
// conversation once.
func (s *IdentitySyncer) RunOnce() {
	if s.merge {
		mergeStats, err := s.client.MergeLIDChats(s.store)
		if err != nil {
			s.client.logger.Warnf("[IDENTITY] LID chat merge failed: %v", err)
		} else if mergeStats.ChatsMerged > 0 {
			s.client.logger.Infof("[IDENTITY] merged %d LID chats (%d messages moved, %d duplicates dropped, %d still unmapped)",
				mergeStats.ChatsMerged, mergeStats.MessagesMoved, mergeStats.Duplicates, mergeStats.Unresolved)
		}
	}

	stats, err := s.client.SyncIdentities(s.store)
	if err != nil {
		s.client.logger.Warnf("[IDENTITY] directory sync failed: %v", err)
		return
	}
	s.client.logger.Infof("[IDENTITY] directory: %d rows (%d users, %d groups, %d lid↔phone linked, %d pruned) in %v",
		stats.Rows, stats.Users, stats.Groups, stats.Linked, stats.Pruned, stats.Duration.Round(time.Millisecond))
}
